
# Predicting the Quality Score of Wine Using Machine Learning
![wines-lined-up](./img/wine-tasting.jpg)
#### Sep 22, 2022
- [Luis Arce]( https://github.com/LuisVArce)
- [Tim Keriazes]( https://github.com/tim-keriazes)
- [Joshua Mayes]( https://github.com/MrEnigmamgine)


In [1]:
# DS Modules
import numpy as np
import pandas as pd

# Visualization modules
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

# Helpers
import helpers
import wrangle
import model
import viz

In [2]:
# Establish some color themes for the notebook to mimic the colors of wine.
wine_colors = ['#efdaa3', '#8c0f0a']
alt_white = '#eccd13'
wine_palette = sns.color_palette(wine_colors)
wine_palette_r = sns.color_palette(list(reversed(wine_colors)))
sns.set_palette(wine_palette)


---
## About this notebook:
This notebook presents a breif overview the key techniques, findings, and acheivements from the project and summarizes the results of our efforts.

### Project Description:
Our project examines 11 quantitative features of red/white wine data sets from the Vihno Verde region of Portugal. Using the physicochemical features/breakdown of the wine, we built a predictive machine learning model with a target variable of quality score. Our insights, discoveries, and modeling offer a distinct advantage to wine producers/stakeholders/distributors by using a wine's chemical composition and predicting its associated quality score.


### Notebook Outline:
<!-- 1. Intro
2. Executive Summary
3. Data Wrangling
4. Split
5. Exploration Highlights
    6. Stats Tests?
    8. Clusters
        7. Scale
9. Modeling
10. Conclusion
11. Next Steps -->




---
## About the data
Our project makes use of the [Wine Quality Data Set](https://archive.ics.uci.edu/ml/datasets/wine+quality) from the [UCI Machine Learning Repository](https://archive.ics.uci.edu/ml/index.php), which is a labeled dataset consisting of 6500 observations.  Each observation represents a red or white Vihno Verde whine from the Portugal region and includes the physicochemical composition of the wine as well as a labeled `quality` score indicating wine expert's opinion of the wine on a scale of 1 to 10.


### Target variable
**`quality`** - Quality score is the median score given to associated wine based off the rankings of three industry experts

<div style="background-color: #f26200; color: white; border-radius: 25px; padding: 15px; width: 55%">
 <strong>Caution!</strong> <br>
  Models trained on biased data will reflect the bias of the humans that labeled it.  <br>
  Because this data was created by a small group, there is a high risk of bias.
</div>

### Data Dictionary
| **Variable Name** | **Explanation** | **Unit** | **Values** |
| :---: | :---: | :---: | :---: |
| Fixed Acidity |  Acids that do not evaporate readily (Tartaric Acid) | g/L | Float |
| Volatile Acidity | Acids evaporate readily (Acetic acid) | g/L | Float |
| Citric Acid | level of Citric acid | g/L | Float |
| Residual Sugar | Sugar that remains after fermenation | g/L | Float |
| Chlorides | Sodium Chloride content | g/L | Float |
| Free Sulfur Dioxide | Levels of free, gaseous sulfur dioxide | mg/L | Float |
| Total Sulfur Dioxide | Total Level of Sulfur Dioxide | mg/L | Float |
| Density | Density in relation to water | g/cm^3 | Float |
| pH| Acidity of the wine | 1-14 | Float |
| Sulphates | Level of potassium sulfate | g/L | Float |
| Alcohol | Alcohol by Volume per wine | ABV% | Float |
| Quality |  The median value of at least 3 independent evualations by wine experts| 1-10 | Integer |



## Data Challenges and preparation
The dataset used was a very clean dataset.  There were zero null values and none of the values could be identified as obviously erroneous.  However, we discovered 1,177 duplicate records which we dropped leaving us with 5320 observations remaining after cleaning, 1359 of which are red wines and 3961 of which are white wines. 

## Outliers
Identifying outliers was a particularly challenging process due to the differences between red and white wines.  Effectively we needed to consider outliers twice, once for each type of wine.  Rather than relying on a cookie-cutter method such as the IQR method, we took a subjective approach and thought critically about which values would be considered normal for a wine of this type.  Handling outliers was particularly helpful during the exploration process because it allowed us to create better charts while taking into account the type of wine the data was coming from.

## Legal Definitions
Researching the legal definitions of wines was also a particularly helpful endeavor.  By identifying the differences between EU and US laws we became more familiar with what is acceptable in a wine.  We found that the most important qualities in the US and EU laws were Sulfur Dioxide and Volatile Acidity.  EU's laws were more restrictive, but had varying limits based on how much residual sugar was in a wine.




#### US
<table>
    <tr>
        <td>
            <strong>Red</strong>
        </td>
        <td>
            <strong>White</strong>
        </td>
    </tr>
    <tr>
        <td>
            SO2 : 350 mg/L
        </td>
        <td>
            SO2 : 350 mg/L
        </td>
    </tr>
    <tr>
        <td>
            VA : 1.4 g/L
        </td>
        <td>
            VA : 1.5 g/L
        </td>
    </tr>
</table>

#### EU
<table>
    <tr>
        <td>
            <strong>Red</strong>
        </td>
        <td>
            <strong>White</strong>
        </td>
    </tr>
    <tr >
        <td>
            SO2 limits by sugar
            <table>
                <tr>
                    <td>
                        sugar &lt; 5g/L
                    </td>
                    <td>
                        150 mg/L
                    </td>
                </tr>
                <tr>
                    <td>
                        5g/L &lt; sugar &lt; 35 g/L
                    </td>
                    <td>
                        250 mg/L
                    </td>
                </tr>
            </table>
        </td>
        <td>
            SO2 limits by sugar:
            <table >
                <tr>
                    <td>
                        sugar &lt; 5g/L
                    </td>
                    <td>
                        200 mg/L
                    </td>
                </tr>
                <tr >
                    <td>
                        5g/L &lt; sugar &lt; 35 g/L
                    </td>
                    <td>
                        250 mg/L
                    </td>
                </tr>
                <tr >
                    <td>
                        35 g/L &lt; sugar
                    </td>
                    <td>
                        400 mg/L
                    </td>
                </tr>
            </table>
        </td>
    </tr>
    <tr >
        <td>
            VA : 1.2 g/L
        </td>
        <td >
            VA : 1.08 g/L
        </td>
    </tr>
</table>

[source 1](https://www.wineshopathome.com/understanding-sulfur-levels-wine/)  
[source 2](https://www.law.cornell.edu/cfr/text/27/4.21)  
[source 3](https://www.mpi.govt.nz/dmsdocument/34896-European-Union-Wine-OMAR)


## Feature Engineering

#### Hydronium
Potential Hydrogen (pH) is an arbitrary scale from 0 to 14 that describes how acidic or basic a solution is.  This is determined by its Hydronium concentration.  We exctacted the hydronium concentration knowing that the formula for pH is $-\log{H_3O{+}}$, in turn the formula used to derive the hydronium concentration is $10^{-pH}$.  We chose to use hydronium instead of pH so that the value was represented in a continuous measure of density instead of a value centered on 7.

#### Bound Sulfur Dioxide
Bound sulfur dioxide is calculated by subtracting the free sulfur dioxide from the total sulfur dioxide.  Free sulfur dioxide describes the amount of sulfur dioxide that is available for bonding, therefore the difference can be inferred as the amount of sulfur dioxide that is already bound and no longer available for bonding.

Because total sulfur dioxide contains the value from free sulfur dioxide there is some multicolinearity introduced by the two variables.  To compensate for that multicolinearity this feature should be calculated and the total sulfur dioxide column should be dropped before modeling.

#### Flavor Profiles
During our research on the variables included in this dataset we found that some variables are considered to have more profound impacts on flavor than others.  By running a K-means clustering model on the most impactful variables, we found 4 disctinct "flavor profiles" that describe the flavor of a wine and highlight the prefences of the wine testers.


### Engineered Feature Dictionary

| **Variable Name** | **Explanation** | **Unit** | **Values** |
| :---: | :---: | :---: | :---: |
| Hydronium | A representation of pH in concentration | M/L | float |
| Bound Sulfur Dioxide | Difference of free sulfur dioxide and total sulfur dioxide | mg/L | float |
| Flavor Profile | K-means cluster using volatile acidity, chlorides, residual sugar, and alcohol | 0-3 | category |

In [3]:
# Load the data into the notebook
train, validate, test = wrangle.wrangle_data_and_split('both')

train = model.drop_outliers(train, 'white')


---
# Exploration Summary
<!-- I feel like something should be said here. -->


## Key Findings
### Regardless of color, a wine with an ABV above 12% is much more likely to be of high quality.


In [4]:
# Draw ABV boxen plot
viz.abv_boxen_by_quality(train)

/Users/luisarce/codeup-data-science/jtl_wine_quality/viz.py:16: UserWarning: Matplotlib is currently using agg, which is a non-GUI backend, so cannot show the figure.
  plt.show()


The the chart above we can see the distribution of Alcohol by Volume (ABV) for each quality.  For each quality score we can find wines as with a low ABV between 8% and 9% as well as a high ABV between 14.5% and 13%.  We can infer that ABV is a great predictor for good wines because the most dense regeions for each category show an upward trend.  

However, the distributions for qualities 3 and 4 tell us a different story.  Because the distributions are quite similar, and in-fact have a higher mean than quality 5 wines, we can infer that there is something else going on that causes a wine to be considered a bad wine.

### Red and white wines are different worlds.

<div style="background-color: #f26200; color: white; border-radius: 25px; padding: 15px; width: 55%">
 <strong>Caution!</strong> <br>
  A model only performs well for the world in which it was trained.
</div>

In [5]:
# Draw plots to show differences between red and white wines.
fig, ax = plt.subplots(2,2,figsize=[10,7])
fig.suptitle('Physicochemical differences between wines', fontsize=16)
sns.histplot(data=train, x='chlorides', hue='type', ax=ax[0][0],alpha=.9)
sns.histplot(data=train, x='residual_sugar', hue='type', ax=ax[0][1],alpha=.8)
sns.histplot(data=train, x='volatile_acidity', hue='type', ax=ax[1][0],alpha=.9)
sns.histplot(data=train, x='total_sulfur_dioxide', hue='type', ax=ax[1][1],alpha=.9)
# Remove most of the legends to declutter the figure.
ax[0][0].get_legend().remove()
ax[1][0].get_legend().remove()
ax[1][1].get_legend().remove()
plt.show()

/var/folders/xr/rf33rdr91vd7xt659vgf8qkm0000gn/T/ipykernel_41096/1752025585.py:12: UserWarning: Matplotlib is currently using agg, which is a non-GUI backend, so cannot show the figure.
  plt.show()


The figure above demonstrates some of the noticeable differences between the types of wines.  It might come as no surprise to learn that red wines and white wines are different types of wines, but this distinction becomes very important when applying machine learning models.  A model is only good in the world it was trained in.  Because of this we have decided that exploring and modeling should be done on each wine type separately.

### Flavor Profiles

In [6]:
# Demonstrate flavor profiles
# Create the profiler instance
profiler = model.FlavorProfile()
# Fit it to the training data
profiler.fit(train)
# Assign clusters to the data
train['flavor_profile'] = profiler.predict(train)
# Draw the plot
plt.figure(figsize=[10,5])
sns.boxplot(data=train, x='flavor_profile', y='quality', hue='type')
plt.title('Quality scores per flavor profile')
plt.show()

/var/folders/xr/rf33rdr91vd7xt659vgf8qkm0000gn/T/ipykernel_41096/543709208.py:12: UserWarning: Matplotlib is currently using agg, which is a non-GUI backend, so cannot show the figure.
  plt.show()


![flavor_profile visualization](./img/plotly_flavor_profiles.png)


By creating these clusters we discovered that the testers for this dataset hat a clear preference for strong, unsweetened wines.  In addition to being a strong predictor for quality, these clusters can help manufacturers market their wine to the appropriate customers.  For example, people who only drink socially might prefer sweeter wines with a low alcohol strength.

## Feature Engineering

### Legality

In [7]:
train['eu_illegal'] = train.apply(viz.not_eu_legal, axis=1)
plt.figure(
    facecolor='white'
    )
plt.pie(train.eu_illegal.value_counts(), explode=(0,0.25), labels=['Legal', 'Not Legal'])
plt.title('Only a small fraction of wines are not legal')
plt.show()

/var/folders/xr/rf33rdr91vd7xt659vgf8qkm0000gn/T/ipykernel_41096/224279765.py:7: UserWarning: Matplotlib is currently using agg, which is a non-GUI backend, so cannot show the figure.
  plt.show()


In [8]:
train[train.eu_illegal].shape, train.shape

((26, 15), (2755, 15))

By testing if each observation's values fall within the limits defined by EU law we engineered the `eu_illegal` column.  We discovered only 26 wines out of the 2755 wines in our training dataset were not EU legal wines.  However we were surprised that the distribution of quality scores for the "illegal" wines looked very similar to the overal distribution of quality scores.  Due to this, we decided not to use this feature in any of our modeling.

In [9]:
# Display quality score distribution for not legal wiens
sns.kdeplot(data=train[train.eu_illegal], x='quality', color=wine_colors[1])
plt.title('Distribution of quality scores for Not Legal wines')
plt.show()

/var/folders/xr/rf33rdr91vd7xt659vgf8qkm0000gn/T/ipykernel_41096/226697335.py:4: UserWarning: Matplotlib is currently using agg, which is a non-GUI backend, so cannot show the figure.
  plt.show()


### Hydronium

In [10]:
#Engineer the feature
train['hydronium'] = 10**(-train['ph'])

Potential Hydrogen (pH) is an arbitrary scale from 0 to 14 that describes how acidic or basic a solution is.  This is determined by its Hydronium concentration.  
We exctacted the hydronium concentration knowing that the formula for pH is $-\log{H_3O{+}}$, in turn the formula used to derive the hydronium concentration is $10^{-pH}$.  
We chose to use hydronium instead of pH so that the value was represented in a continuous measure of density instead of a value centered on 7.

In [11]:
# Show the relationship between pH and Hydronium
sns.set_palette(wine_palette_r)
sns.lineplot(data=train, x='ph', y='hydronium')
plt.title('Polynomial relationship between pH and Hydronium')
plt.show()

/var/folders/xr/rf33rdr91vd7xt659vgf8qkm0000gn/T/ipykernel_41096/1545469460.py:5: UserWarning: Matplotlib is currently using agg, which is a non-GUI backend, so cannot show the figure.
  plt.show()


Here we can see how the transformation changes the shape of the data.

In [12]:
# Show comparison plot between pH and Hydronium
sns.set_palette(wine_palette)
fig, axes = plt.subplots(1,2, figsize=[10,4])
sns.barplot(data=train, x='quality', y='hydronium', hue='type', ax=axes[0])
sns.barplot(data=train, x='quality', y='ph', hue='type', ax=axes[1])
fig.suptitle('Comparing pH to Hydronium against quality')
for ax in axes:
    ax.get_legend().remove()
plt.show()

/var/folders/xr/rf33rdr91vd7xt659vgf8qkm0000gn/T/ipykernel_41096/482441308.py:9: UserWarning: Matplotlib is currently using agg, which is a non-GUI backend, so cannot show the figure.
  plt.show()


The above comparison demonstrates how transforming pH into Hydronium concentration can be more predictive of quality than pH alone.

## Most predictive qualities

Throughout our exploration and testing, we found the following features to be the most predictive of a wine's quality for both red and white wines.
- alcohol
- volatile_acidity
- citric_acid
- residual_sugar
- total_sulfur_dioxide
- hydronium
- density
- ions

In [13]:
# Store the features list for later.
feats = [
    'alcohol',
    'volatile_acidity',
    'citric_acid',
    'residual_sugar',
    'total_sulfur_dioxide',
    'hydronium',
    'density',
    'ions',
    ]

# Ensure features exist in dataset
train = model.add_features(train)

# A counter for iterating through feats
i = 0

Even though the list of features is consistent across both wine colors, each one behaves differently for each color.

In [14]:
# Compare ABV
viz.regplot_compare(train, 'quality', feats[i])
i += 1

/Users/luisarce/codeup-data-science/jtl_wine_quality/viz.py:62: UserWarning: Matplotlib is currently using agg, which is a non-GUI backend, so cannot show the figure.
  plt.show()


- Red and white wines behave similarly here, as the % Alcohol by Volume increases, so does the average quality score.

In [15]:
# Compare Volatile acidity
viz.regplot_compare(train, 'quality', feats[i])
i += 1

/Users/luisarce/codeup-data-science/jtl_wine_quality/viz.py:62: UserWarning: Matplotlib is currently using agg, which is a non-GUI backend, so cannot show the figure.
  plt.show()


- When comparing the effect of volatile acidity on a wine's quality, the need to model the wines seperately becomes even more clear.  Although both colors show a negative correllation to quality, there is a clear difference in the patterns.  A low quality white wine with a volatile acidity of 0.4 could be seen as similar to a high quality red wine with a volatile acidity of 0.4.

In [16]:
# Compare Citric Acid
viz.regplot_compare(train, 'quality', feats[i])
i += 1

/Users/luisarce/codeup-data-science/jtl_wine_quality/viz.py:62: UserWarning: Matplotlib is currently using agg, which is a non-GUI backend, so cannot show the figure.
  plt.show()


- The concentration of citric acid tells a similar story and highlights the need to separate models.

In [17]:
# Compare Residual Sugar
viz.regplot_compare(train, 'quality', feats[i])
i += 1

/Users/luisarce/codeup-data-science/jtl_wine_quality/viz.py:62: UserWarning: Matplotlib is currently using agg, which is a non-GUI backend, so cannot show the figure.
  plt.show()


- Comparing residual sugar content also highlights the need for separate models.  We can see a much more dramatic variance in residual sugar content among the white wines that yeilds some predictive power while the residual sugar content of red wines seems to have no effect on quality 
- The non-linear pattern suggests that using polynomal features may be beneficial.

In [18]:
# Compare Total sulfur dioxide
viz.regplot_compare(train, 'quality', feats[i])
i += 1

/Users/luisarce/codeup-data-science/jtl_wine_quality/viz.py:62: UserWarning: Matplotlib is currently using agg, which is a non-GUI backend, so cannot show the figure.
  plt.show()


- Total sulfur dioxide content shows a similar story to the residual sugar content where the range differences in colors are important to keep separate.

In [19]:
# Compare Hydronium
viz.regplot_compare(train, 'quality', feats[i])
i += 1

/Users/luisarce/codeup-data-science/jtl_wine_quality/viz.py:62: UserWarning: Matplotlib is currently using agg, which is a non-GUI backend, so cannot show the figure.
  plt.show()


- Comparing Hydronium concentrations highlights perhaps one of the most stark contrasts between red and white wines.  It appears that white wines become less acidic as they increase in quality whereas red wines become *more* acidic.

In [20]:
# Compare Density
viz.regplot_compare(train, 'quality', feats[i])
i += 1

/Users/luisarce/codeup-data-science/jtl_wine_quality/viz.py:62: UserWarning: Matplotlib is currently using agg, which is a non-GUI backend, so cannot show the figure.
  plt.show()


- Density seems to have a much more dramatic corellation with quality for white wines.

In [21]:
# Compare Ions
viz.regplot_compare(train, 'quality', feats[i])
i += 1

/Users/luisarce/codeup-data-science/jtl_wine_quality/viz.py:62: UserWarning: Matplotlib is currently using agg, which is a non-GUI backend, so cannot show the figure.
  plt.show()


- Ions seem to be much more important for red wines than they are for white wines.

---
# Modeling

In [22]:
# Libraries used for modeling
from sklearn.preprocessing import PolynomialFeatures
from sklearn.metrics import mean_squared_error

from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor
from sklearn.neighbors import KNeighborsRegressor

In this notebook we will be demonstrating the performance of our best models compared to a simple baseline.

### Preparing the data for modeling

In [23]:
# First, declare the variables we will be feeding into the models.
target = 'quality'
feats = [
    'alcohol',
    'volatile_acidity',
    'citric_acid',
    'residual_sugar',
    'total_sulfur_dioxide',
    'hydronium',
    'density',
    'ions',
    ]

In [24]:
# Prepare the dataset for modeling

# Prepare the red wine samples
red = wrangle.wrangle_data("red")
red = model.add_features(red)
red_train, red_validate, red_test = wrangle.train_test_validate_split(red)
red_train = model.drop_outliers(red_train, "red", method='manual')
red_train, red_validate, red_test = [df[feats + [target]] for df in [red_train, red_validate, red_test]]

# Prepare the white wine samples
white = wrangle.wrangle_data("white")
white = model.add_features(white)
white_train, white_validate, white_test = wrangle.train_test_validate_split(white)
white_train = model.drop_outliers(white_train, "white", method='manual')
white_train, white_validate, white_test = [df[feats + [target]] for df in [white_train, white_validate, white_test]]


In [25]:
# Define the X and y sets used for red models
r_xt = red_train[feats]
r_xv = red_validate[feats]
r_xs = red_test[feats]
r_yt = red_train[target]
r_yv = red_validate[target]
r_ys = red_test[target]

red_scaler = helpers.prep.train_scaler(r_xt)
r_xt = red_scaler.transform(r_xt)
r_xv = red_scaler.transform(r_xv)
r_xs = red_scaler.transform(r_xs)

red_poly = PolynomialFeatures(2)
red_poly.fit(r_xt)
r_xt_poly = red_poly.transform(r_xt)
r_xv_poly = red_poly.transform(r_xv)
r_xs_poly = red_poly.transform(r_xs)

In [26]:
# Define the X and y sets used for white models
w_xt = white_train[feats]
w_xv = white_validate[feats]
w_xs = white_test[feats]
w_yt = white_train[target]
w_yv = white_validate[target]
w_ys = white_test[target]

white_scaler = helpers.prep.train_scaler(w_xt)
w_xt = white_scaler.transform(w_xt)
w_xv = white_scaler.transform(w_xv)
w_xs = white_scaler.transform(w_xs)

white_poly = PolynomialFeatures(3)
white_poly.fit(w_xt)
w_xt_poly = white_poly.transform(w_xt)
w_xv_poly = white_poly.transform(w_xv)
w_xs_poly = white_poly.transform(w_xs)

## Red Models

In [27]:
# Declare and fit models
red_models = {
    'Mean Baseline': helpers.eval.BaselineModel(method='mean'),
    'Linear Regression': LinearRegression(positive=True),
    'Random Forest': RandomForestRegressor(max_depth=10),
    'KNN Regression': KNeighborsRegressor(n_neighbors=5),
}
red_poly_models = {
    'Linear Regression': LinearRegression(positive=True),
    'Random Forest': RandomForestRegressor(max_depth=10),
    'KNN Regression': KNeighborsRegressor(n_neighbors=5),
}

for name, model in red_models.items():
    model.fit(r_xt, r_yt)
for name, model in red_poly_models.items():
    model.fit(r_xt_poly, r_yt)


In [28]:
# Collect scores for red models
scores = {}
for name, model in red_models.items():
    tp = model.predict(r_xt)
    vp = model.predict(r_xv)
    scores[f'{name}'] = {
        'type' : 'red',
        'train_rmse' : mean_squared_error(r_yt, tp, squared=False),
        'validate_rmse' : mean_squared_error(r_yv, vp, squared=False),
        'tp': tp,
        'vp': vp
    }

for name, model in red_poly_models.items():
    tp = model.predict(r_xt_poly)
    vp = model.predict(r_xv_poly)
    scores[f'{name} (Polynomial)'] = {
        'type' : 'red',
        'train_rmse' : mean_squared_error(r_yt, tp, squared=False),
        'validate_rmse' : mean_squared_error(r_yv, vp, squared=False),
        'tp': tp,
        'vp': vp
    }

# Store scores in a dataframe
scores_df = pd.DataFrame.from_dict(scores, orient='index')

In [29]:
scores_df

type  train_rmse  validate_rmse  \
Mean Baseline                   red    0.838951       0.803947   
Linear Regression               red    0.700431       0.701773   
Random Forest                   red    0.342062       0.617276   
KNN Regression                  red    0.588769       0.682125   
Linear Regression (Polynomial)  red    0.698293       0.683348   
Random Forest (Polynomial)      red    0.341163       0.619635   
KNN Regression (Polynomial)     red    0.586810       0.684599   

                                                                               tp  \
Mean Baseline                   [5.641361256544503, 5.641361256544503, 5.64136...   
Linear Regression               [5.181408708074526, 5.600099971704844, 5.12615...   
Random Forest                   [4.252000015547554, 5.781321001617499, 5.15182...   
KNN Regression                  [4.6, 5.2, 5.0, 5.4, 5.6, 5.0, 5.8, 5.0, 6.2, ...   
Linear Regression (Polynomial)  [5.203534593143311, 5.5924258553601, 5.1617620...   
Random Forest (Polynomial)      [4.3185090933103645, 5.717676927545992, 5.1226...   
KNN Regression (Polynomial)     [4.6, 5.2, 5.0, 5.4, 5.6, 5.0, 5.8, 5.4, 6.2, ...   

                                                                               vp  
Mean Baseline                   [5.641361256544503, 5.641361256544503, 5.64136...  
Linear Regression               [5.09168456875291, 5.4949158559919775, 5.29490...  
Random Forest                   [5.250529443020627, 5.659239868026149, 5.48402...  
KNN Regression                  [5.2, 5.6, 5.2, 5.4, 5.2, 5.4, 4.8, 6.2, 5.2, ...  
Linear Regression (Polynomial)  [5.121456620599672, 5.491387735632516, 5.30863...  
Random Forest (Polynomial)      [5.150091945902969, 5.622874233294231, 5.52090...  
KNN Regression (Polynomial)     [5.0, 5.6, 5.4, 5.4, 5.2, 5.4, 4.8, 6.2, 5.4, ...

## White Models

In [30]:
# Declare and fit models
white_models = {
    'Mean Baseline': helpers.eval.BaselineModel(method='mean'),
    'Linear Regression': LinearRegression(positive=True),
    'Random Forest': RandomForestRegressor(max_depth=10),
    'KNN Regression': KNeighborsRegressor(n_neighbors=5),
}
white_poly_models = {
    'Linear Regression': LinearRegression(positive=True),
    'Random Forest': RandomForestRegressor(max_depth=10),
    'KNN Regression': KNeighborsRegressor(n_neighbors=5),
}
for name, model in white_models.items():
    model.fit(w_xt, w_yt)
for name, model in white_poly_models.items():
    model.fit(w_xt_poly, w_yt)

In [31]:
# Collect scores for white models
scores = {}
for name, model in white_models.items():
    tp = model.predict(w_xt)
    vp = model.predict(w_xv)
    scores[f'{name}'] = {
        'type' : 'white',
        'train_rmse' : mean_squared_error(w_yt, tp, squared=False),
        'validate_rmse' : mean_squared_error(w_yv, vp, squared=False),
        'tp': tp,
        'vp': vp
    }

for name, model in white_poly_models.items():
    tp = model.predict(w_xt_poly)
    vp = model.predict(w_xv_poly)
    scores[f'{name} (Polynomial)'] = {
        'type' : 'white',
        'train_rmse' : mean_squared_error(w_yt, tp, squared=False),
        'validate_rmse' : mean_squared_error(w_yv, vp, squared=False),
        'tp': tp,
        'vp': vp
    }

# Store them in a dataframe
temp = pd.DataFrame.from_dict(scores, orient='index')

In [32]:
temp

type  train_rmse  validate_rmse  \
Mean Baseline                   white    0.875574       0.926019   
Linear Regression               white    0.767971       0.819626   
Random Forest                   white    0.469100       0.753727   
KNN Regression                  white    0.629194       0.805003   
Linear Regression (Polynomial)  white    0.766012       0.819692   
Random Forest (Polynomial)      white    0.419829       0.752931   
KNN Regression (Polynomial)     white    0.633970       0.801577   

                                                                               tp  \
Mean Baseline                   [5.912942271880819, 5.912942271880819, 5.91294...   
Linear Regression               [5.760700277856108, 5.802331934610655, 6.42675...   
Random Forest                   [5.80177584925685, 5.939411996741085, 6.888362...   
KNN Regression                  [5.8, 5.8, 6.6, 6.8, 6.0, 6.8, 5.2, 6.6, 5.6, ...   
Linear Regression (Polynomial)  [5.769014193911812, 5.794248526453551, 6.48406...   
Random Forest (Polynomial)      [5.778505972898781, 5.922750245979631, 7.00845...   
KNN Regression (Polynomial)     [5.8, 5.8, 6.0, 6.8, 6.0, 6.8, 5.2, 6.6, 5.6, ...   

                                                                               vp  
Mean Baseline                   [5.912942271880819, 5.912942271880819, 5.91294...  
Linear Regression               [5.840623970700581, 5.492898184016904, 5.39559...  
Random Forest                   [5.69224005600053, 5.009639240035029, 5.978087...  
KNN Regression                  [6.0, 5.8, 5.4, 5.2, 5.0, 6.0, 5.2, 6.2, 6.6, ...  
Linear Regression (Polynomial)  [5.834425981644936, 5.4904627431583855, 5.4134...  
Random Forest (Polynomial)      [5.825970721656278, 4.973231488052765, 5.85586...  
KNN Regression (Polynomial)     [6.2, 5.6, 5.4, 5.4, 5.0, 6.4, 5.2, 6.0, 6.4, ...

## Comparing the models

In [33]:
# Concatonate scores into same dataframe
scores_df = pd.concat([scores_df, temp])
scores_df = scores_df.reset_index().rename(columns={'index': 'model'})

In [36]:

# And plot
plt.figure(figsize=[7,5])
ax = sns.barplot(data=scores_df, y='model', x='validate_rmse', hue='type', palette=wine_palette_r)
ax.get_legend().remove()
plt.title('Best models')
plt.show()

/var/folders/xr/rf33rdr91vd7xt659vgf8qkm0000gn/T/ipykernel_41096/1477798376.py:6: UserWarning: Matplotlib is currently using agg, which is a non-GUI backend, so cannot show the figure.
  plt.show()
